### PR TITLE
EOS-16257: Build s3confstore and run UTs with jenkins build job

### PR DESCRIPTION
**Pull request by pratyush-seagate**
commit d457679cf58f716352cd90c3aaef5f182b06c3c4
Author: pratyush-seagate <pratyush.k.khan@seagate.com>
Date:   Mon Jan 11 00:41:41 2021 -0700

        EOS-16257: Build s3confstore and run UTs with jenkins build job
            Change description:
                Added build step in rebuildall.sh script (just like s3recovery and s3backgrounddelete) for s3confstore.
                Added UT for s3confstore and made sure they run as part of jenkins job via runalltest.sh
            List of files:
                modified:   rebuildall.sh
                modified:   runalltest.sh
                new file:   s3cortxutils/s3confstore/scripts/run_all_ut.sh
                new file:   s3cortxutils/s3confstore/ut/s3confstore_unit_tests.py
            Testing:
            UT:
                    1. using rebuildall.sh, using console logs, verified s3confstore is being compiled.
                2. using ./jenkins-build.sh and ./runalltest.sh, verified s3confstore_unit_tests.py is being executed successfully.
                    Note: pip3 install  ./cortx_py_utils-1.0.0-py3-none-any.whl needed to install cortx-py-utils before running s3confstore unit test.

    Signed-off-by: pratyush-seagate <pratyush.k.khan@seagate.com>

### CODE REVIEW

**Comment by pratyush-seagate** [comment]
From console build logs:

/root/cortx-s3server
++ '[' 0 -eq 0 ']'
++ cd s3cortxutils/s3confstore
++ '[' 0 -eq 0 ']'
++ python36 setup.py install --force
running install
running bdist_egg
running egg_info
creating s3confstore.egg-info
writing s3confstore.egg-info/PKG-INFO
writing dependency_links to s3confstore.egg-info/dependency_links.txt
writing top-level names to s3confstore.egg-info/top_level.txt
writing manifest file 's3confstore.egg-info/SOURCES.txt'
reading manifest file 's3confstore.egg-info/SOURCES.txt'
writing manifest file 's3confstore.egg-info/SOURCES.txt'
installing library code to build/bdist.linux-x86_64/egg
running install_lib
running build_py
creating build
creating build/lib
creating build/lib/s3confstore
copying s3confstore/cortx_s3_confstore.py -> build/lib/s3confstore
copying s3confstore/__init__.py -> build/lib/s3confstore
copying s3confstore/s3confstore -> build/lib/s3confstore
creating build/bdist.linux-x86_64
creating build/bdist.linux-x86_64/egg
creating build/bdist.linux-x86_64/egg/s3confstore
copying build/lib/s3confstore/cortx_s3_confstore.py -> build/bdist.linux-x86_64/egg/s3confstore
copying build/lib/s3confstore/s3confstore -> build/bdist.linux-x86_64/egg/s3confstore
copying build/lib/s3confstore/__init__.py -> build/bdist.linux-x86_64/egg/s3confstore
byte-compiling build/bdist.linux-x86_64/egg/s3confstore/cortx_s3_confstore.py to cortx_s3_confstore.cpython-36.pyc
byte-compiling build/bdist.linux-x86_64/egg/s3confstore/__init__.py to __init__.cpython-36.pyc
creating build/bdist.linux-x86_64/egg/EGG-INFO
installing scripts to build/bdist.linux-x86_64/egg/EGG-INFO/scripts
running install_scripts
running build_scripts
creating build/scripts-3.6
copying and adjusting s3confstore/s3confstore -> build/scripts-3.6
changing mode of build/scripts-3.6/s3confstore from 644 to 755
creating build/bdist.linux-x86_64/egg/EGG-INFO/scripts
copying build/scripts-3.6/s3confstore -> build/bdist.linux-x86_64/egg/EGG-INFO/scripts
changing mode of build/bdist.linux-x86_64/egg/EGG-INFO/scripts/s3confstore to 755
copying s3confstore.egg-info/PKG-INFO -> build/bdist.linux-x86_64/egg/EGG-INFO
copying s3confstore.egg-info/SOURCES.txt -> build/bdist.linux-x86_64/egg/EGG-INFO
copying s3confstore.egg-info/dependency_links.txt -> build/bdist.linux-x86_64/egg/EGG-INFO
copying s3confstore.egg-info/top_level.txt -> build/bdist.linux-x86_64/egg/EGG-INFO
zip_safe flag not set; analyzing archive contents...
creating dist
creating 'dist/s3confstore-1.0.0-py3.6.egg' and adding 'build/bdist.linux-x86_64/egg' to it
removing 'build/bdist.linux-x86_64/egg' (and everything under it)
Processing s3confstore-1.0.0-py3.6.egg
Copying s3confstore-1.0.0-py3.6.egg to /usr/local/lib/python3.6/site-packages
Adding s3confstore 1.0.0 to easy-install.pth file
Installing s3confstore script to /usr/local/bin

Installed /usr/local/lib/python3.6/site-packages/s3confstore-1.0.0-py3.6.egg
Processing dependencies for s3confstore==1.0.0
Finished processing dependencies for s3confstore==1.0.0
++ cd -

**Comment by pratyush-seagate** [comment]
From Ut run logs on console:

# s3cortxutils/s3confstore/scripts/run_all_ut.sh

Running s3confstore tool UT's...
=================================================================== test session starts ====================================================================
platform linux -- Python 3.6.8, pytest-2.9.2, py-1.4.32, pluggy-0.3.1 -- /bin/python36
cachedir: s3cortxutils/s3confstore/.cache
rootdir: /root/cortx-s3server/s3cortxutils/s3confstore, inifile:
collected 1 items

s3cortxutils/s3confstore/ut/s3confstore_unit_tests.py::test_uttest PASSED

================================================================= 1 passed in 0.04 seconds =================================================================
s3confstore tool UT's runs successfully

**Comment by pratyush-seagate** [comment]
jenkins jobs yet to be run. Local runs are still ongoing.

**Comment by pratyush-seagate** [comment]
> 
> 
> Jenkins job started at http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/S3server-7.8.2003/394/

Update:
Due to missing toml package, s3confstore ut failed. This package is needed because of py-utils's confstore.

14:15:37  ========================== 26 passed in 0.68 seconds ===========================
14:15:37  s3recovery tool UT's runs successfully
14:15:37  
14:15:37  Check s3conftoolut...OK 
14:15:37  
14:15:37  Running s3confstore tool UT's...
14:15:38  ============================= test session starts ==============================
14:15:38  platform linux -- Python 3.6.8, pytest-2.9.2, py-1.4.32, pluggy-0.3.1 -- /usr/bin/python36
14:15:38  cachedir: s3cortxutils/s3confstore/.cache
14:15:38  rootdir: /var/jenkins/workspace/Cortx-PR-Build/S3server-7.8.2003@2/s3cortxutils/s3confstore, inifile: 
14:15:38  collecting ... collected 0 items / 1 errors
14:15:38  
14:15:38  ==================================== ERRORS ====================================
14:15:38  ________________ ERROR collecting ut/s3confstore_unit_tests.py _________________
14:15:38  s3cortxutils/s3confstore/ut/s3confstore_unit_tests.py:22: in <module>
14:15:38      from s3confstore.cortx_s3_confstore import S3CortxConfStore
14:15:38  s3cortxutils/s3confstore/s3confstore/cortx_s3_confstore.py:26: in <module>
14:15:38      from cortx.utils.kv_store import kv_store_collection
14:15:38  /usr/lib/python3.6/site-packages/cortx/utils/kv_store/kv_store_collection.py:22: in <module>
14:15:38      import toml
14:15:38  E   ModuleNotFoundError: No module named 'toml'
14:15:38  =========================== 1 error in 0.07 seconds ============================
14:15:38  
14:15:38  ***************
14:15:38  *** FAILED ***
14:15:38  ***************
14:15:38  
14:15:38  Error encountered. Exiting unit test runs prematurely...
14:15:38  
14:15:38  ***************
14:15:38  *** ABORTED ***
14:15:38  ***************
14:15:38  
14:15:38  Error encountered. Exiting test runs prematurely...
14:15:38  + echo 'S3 Tests failed.'

@amitwac2608  @AjinkyaDhumal  Please advise the desired steps to fix this.

**Comment by pratyush-seagate** [comment]
I've reached out to @vimalraghwani20 on how to add the toml package in init.sh. Once he shares the steps will add it and restart jenkins.

File: runalltest.sh [hunk 1]
```diff
@@ -157,6 +158,12 @@ then
   printf "OK \n"
 
   $UT_S3RECOVERYTOOL 2>&1
+
+  printf "\nCheck s3conftoolut..."
```
Comment (amitwac2608) [inline runalltest.sh:162]: Check s3confstoreut

File: runalltest.sh [hunk 2]
```diff
@@ -121,6 +121,7 @@ then
   UT_MEMPOOLMGR_BIN=./bazel-bin/s3mempoolmgrut
   UT_S3BACKGROUNDDELETE=./s3backgrounddelete/scripts/run_all_ut.sh
   UT_S3RECOVERYTOOL=./s3recovery/scripts/run_all_ut.sh
+  UT_S3CONFTOOL=./s3cortxutils/s3confstore/scripts/run_all_ut.sh
```
Comment (amitwac2608) [inline runalltest.sh:124]: UT_S3CONFSTORE

File: s3cortxutils/s3confstore/scripts/run_all_ut.sh [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+# For any questions about this software or licensing,
+# please email opensource@seagate.com or cortx-questions@seagate.com.
+#
+
+#Script to run UT's of s3recovery tool
```
Comment (amitwac2608) [inline s3cortxutils/s3confstore/scripts/run_all_ut.sh:20]: copy-paste error

File: s3cortxutils/s3confstore/scripts/run_all_ut.sh [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+*** FAILED ***
+***************
+'
+    echo "Error encountered. Exiting unit test runs prematurely..." >&2
+    trap : 0
+    exit 1
+}
+trap 'abort' 0
+
+printf "\nRunning s3confstore tool UT's...\n"
```
Comment (amitwac2608) [inline s3cortxutils/s3confstore/scripts/run_all_ut.sh:36]: remove tool from the statement.

File: s3cortxutils/s3confstore/scripts/run_all_ut.sh [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+
+SCRIPT_PATH=$(readlink -f "$0")
+SCRIPT_DIR=$(dirname "$SCRIPT_PATH")
+
+#Update python path to source modules and run unit tests.
+PYTHONPATH=${PYTHONPATH}:${SCRIPT_DIR}/.. python36 -m pytest $SCRIPT_DIR/../ut/*.py -v
+
+echo "s3confstore tool UT's runs successfully"
```
Comment (amitwac2608) [inline s3cortxutils/s3confstore/scripts/run_all_ut.sh:44]: do not use tool word

File: s3cortxutils/s3confstore/ut/s3confstore_unit_tests.py [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+# please email opensource@seagate.com or cortx-questions@seagate.com.
+#
+
+#!/usr/bin/python3.6
+
+from s3confstore.cortx_s3_confstore import S3CortxConfStore
+
+def test_uttest():
+  s3confstore = S3CortxConfStore()
+  s3confstore.uttest("ut_index")
```
Comment (amitwac2608) [inline s3cortxutils/s3confstore/ut/s3confstore_unit_tests.py:26]: Pratyush - I guess this is not right way to right UTs... can you please refer s3recoverytool UT as suggested and write proper UT ?

**Review by amitwac2608 — CHANGES_REQUESTED** [review]

File: s3cortxutils/s3confstore/ut/s3confstore_unit_tests.py (hunk 6 above)
Comment (pratyush-seagate) [reply s3cortxutils/s3confstore/ut/s3confstore_unit_tests.py:26]: The current s3confstore.uttest() is the inbuilt UT written before, it exercises all the APIs. But ok, I'll also add the individual APIs separately in this UT.

Comment (amitwac2608) [reply s3cortxutils/s3confstore/ut/s3confstore_unit_tests.py:26]: Point is not to invoke all APIs, but to test each entry-exit points of a sub-module, where the python test module would help, by mocking unwanted outside stuff, and just test the sub-module of our class.

**Comment by pratyush-seagate** [comment]
Latest jenkins job passed with this change. ref. http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/S3server-7.8.2003/397/

toml package addition changes worked as suggested by @vimalraghwani20 
From http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/S3server-7.8.2003/397/execution/node/26/log/?consoleFull

01:33:51  Installing toml
01:33:53  WARNING: Running pip install with root privileges is generally not a good idea. Try `pip3 install --user` instead.
01:33:53  Collecting toml
01:33:53    Downloading https://files.pythonhosted.org/packages/44/6f/7120676b6d73228c96e17f1f794d8ab046fc910d781c8d151120c3f1569e/toml-0.10.2-py2.py3-none-any.whl
01:33:53  Installing collected packages: toml
01:33:53  Successfully installed toml-0.10.2



UTs ran successfully:
02:19:14  s3recovery tool UT's runs successfully
02:19:14  
02:19:14  Check s3confstoreut...OK 
02:19:14  
02:19:14  Running s3confstore UT's...
02:19:14  ============================= test session starts ==============================
02:19:14  platform linux -- Python 3.6.8, pytest-2.9.2, py-1.4.32, pluggy-0.3.1 -- /usr/bin/python36
02:19:14  cachedir: s3cortxutils/s3confstore/.cache
02:19:14  rootdir: /var/jenkins/workspace/Cortx-PR-Build/S3server-7.8.2003@2/s3cortxutils/s3confstore, inifile: 
02:19:14  collecting ... collected 3 items
02:19:14  
02:19:14  s3cortxutils/s3confstore/ut/s3confstore_unit_tests.py::S3ConfStoreAPIsUT::test_json_conf PASSED
02:19:14  s3cortxutils/s3confstore/ut/s3confstore_unit_tests.py::S3ConfStoreAPIsUT::test_mock_get PASSED
02:19:14  s3cortxutils/s3confstore/ut/s3confstore_unit_tests.py::S3ConfStoreAPIsUT::test_yaml_conf PASSED
02:19:14  
02:19:14  =========================== 3 passed in 0.10 seconds ===========================
02:19:14  s3confstore UT's runs successfully

**Comment by pratyush-seagate** [comment]
latest jenkins passed: http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/S3server-7.8.2003/398/

**Review by AjinkyaDhumal — APPROVED** [review]

File: s3cortxutils/s3confstore/ut/s3confstore_unit_tests.py [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+import unittest
+import json
+import os.path
+from s3confstore.cortx_s3_confstore import S3CortxConfStore
+from cortx.utils.conf_store import Conf
+
+class S3ConfStoreAPIsUT(unittest.TestCase):
+  """UT to test current APIs"""
+
+  def test_json_conf(self):
```
Comment (amitwac2608) [inline s3cortxutils/s3confstore/ut/s3confstore_unit_tests.py:32]: also test empty config path argument to load_config function.

Now we can remove uttest method from the source python class.

Also write UT for set config

File: s3cortxutils/s3confstore/ut/s3confstore_unit_tests.py [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+      os.remove("/tmp/cortx_s3_confstoreuttest.yaml")
+      self.assertFalse(True)
+    s3confstore.set_config(index, 'bridge>port', '1234', False)
+    result_data = s3confstore.get_config(index, 'bridge>port')
+    if result_data != '1234':
+      os.remove("/tmp/cortx_s3_confstoreuttest.yaml")
+      self.assertFalse(True)
+    os.remove("/tmp/cortx_s3_confstoreuttest.yaml")
+
+  @mock.patch.object(Conf, 'get')
```
Comment (amitwac2608) [inline s3cortxutils/s3confstore/ut/s3confstore_unit_tests.py:73]: we need to write more UTs for get_config method, like when index is empty string , key is empty or invalid  etc.

Comment (amitwac2608) [reply s3cortxutils/s3confstore/ut/s3confstore_unit_tests.py:73]: Idea of UT is to test every input and output to a method.

File: scripts/env/dev/init.sh [hunk 9]
```diff
@@ -103,6 +110,8 @@ else
               ;;
           s)
              source ${S3_SRC_DIR}/scripts/env/common/setup-yum-repos.sh
+             # install pip3 package toml if not present, needed by cortx-py-utils for s3confstore UT
```
Comment (sachitanands) [inline scripts/env/dev/init.sh:113]: Please update same changes to scripts/env/release/init.sh and scripts/env/rpmbuild/init.sh as well

**Comment by pratyush-seagate** [comment]
The last set of changes address all the review comments from @amitwac2608  and @sachitanands and fixes codacy warnings. This also rebases mini-provisioner head. Latest jenkins job running at http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/S3server-7.8.2003/408/

**Review by sachitanands — APPROVED** [review]
looks good to me ..
please address Amit kumar's review comments

**Comment by pratyush-seagate** [comment]
> 
> 
> The last set of changes address all the review comments from @amitwac2608 and @sachitanands and fixes codacy warnings. This also rebases mini-provisioner head. Latest jenkins job running at http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/S3server-7.8.2003/408/

Update on jenkins:
jenkins jobs are failing on this branch as well as in the mini-provisioner branch due to the error:
14:07:31  + sh /var/jenkins/workspace/Cortx-PR-Build/S3server-7.8.2003@2/scripts/create_auth_jks_password.sh
14:07:31  Generating password for jks keystore used in authserver...
14:07:31  ERROR: failed to generate key using py-utils:Cipher utility, exiting.
[Pipeline] }
[Pipeline] // script
[Pipeline] }
[Pipeline] // stage
[Pipeline] stage
[Pipeline] { (S3 RPM build)
Stage "S3 RPM build" skipped due to earlier failure(s)

This branch was recently rebased to mini-provisioner head and hence the same failure is being seen here as well. 

mini-provisioner jenkins failure:
http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/S3server-7.8.2003/411/console

**Review by amitwac2608 — APPROVED** [review]

**Comment by pratyush-seagate** [comment]
> 
> 
> > The last set of changes address all the review comments from @amitwac2608 and @sachitanands and fixes codacy warnings. This also rebases mini-provisioner head. Latest jenkins job running at http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/S3server-7.8.2003/408/
> 
> Update on jenkins:
> jenkins jobs are failing on this branch as well as in the mini-provisioner branch due to the error:
> 14:07:31 + sh /var/jenkins/workspace/Cortx-PR-Build/S3server-7.8.2003@2/scripts/create_auth_jks_password.sh
> 14:07:31 Generating password for jks keystore used in authserver...
> 14:07:31 ERROR: failed to generate key using py-utils:Cipher utility, exiting.
> [Pipeline] }
> [Pipeline] // script
> [Pipeline] }
> [Pipeline] // stage
> [Pipeline] stage
> [Pipeline] { (S3 RPM build)
> Stage "S3 RPM build" skipped due to earlier failure(s)
> 
> This branch was recently rebased to mini-provisioner head and hence the same failure is being seen here as well.
> 
> mini-provisioner jenkins failure:
> http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/S3server-7.8.2003/411/console

Update:
@amitwac2608 found the root cause of this problem as jenkins VM's having incorrect cortx-py-utils installed on them. As a workaround, I installed correct cortx-py-utils rpm built from main branch on the jenkins VM for run http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/S3server-7.8.2003/415/

Jenkins job still running and progressed from the last failure point at "sh /var/jenkins/workspace/Cortx-PR-Build/S3server-7.8.2003@2/scripts/create_auth_jks_password.sh"

**Comment by pratyush-seagate** [comment]
latest jenkins passed: http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/S3server-7.8.2003/415/
